text/0005-is-operator.md

```diff
@@ -0,0 +1,176 @@
+- Start Date: 2023-06-16
+- Target Major Version: X.x
+- Reference Issues: <https://github.com/cedar-policy/cedar/issues/94>
+- Implementation PR:
+
+# Summary
+
+This RFC proposes to add an `is` operator to the Cedar language to allow checking equality of entity types.
+
+# Basic example
+
+Say that a user wants to allow any `User` to view files in the folder `Folder::"public"`.
+With an `is` operator, the following policy achieves this goal.
+
+```
+permit(
+  principal is User,
+  action == Action::"view",
+  resource is File in Folder::"public"
+);
+```
+
+At evaluation time, `principal is User` will evaluate to true or false depending on the type of `principal`, and `resource is File in Folder::"public"` will evaluate to true or false depending on the type of `resource`, and whether the `resource` is a member of `Folder::"public"`.
+
+# Motivation
+
+The `is` operator in useful in (at least) two common scenarios.
+
+## Use case 1: Input validation
+
+Say that a user wants to allow any `User` to view files in the folder `Folder::"public"`.
+Currently, to achieve the desired behavior, the user should define a schema with the action entity "view" that can apply to principals of type `User` and resources of type `File`, and then write the following policy.
+
+```
+permit(principal, action == Action::"view", resource in Folder::"public");
+```
+
+Assuming that they pass in inputs where the principal is of type `User` and the resource is of type `File`, this will work as expected.
+However, it could also return "allow" for other principal types like `Folder` or resource types like `User`.
+This is a bug on the user's end: they shouldn’t make a query with the action "view" when the principal is not a `User` and the resource is not a `File`.
+They are breaking the contract they stated in their schema.
+Nevertheless, it might be confusing that such a query does not result in a "deny".
```

Comment:
Do we have an example of using `is` that is not defending against erroneous use of the schema? Maybe we can rework the example to do that? I'd feel more persuaded by that then using redundant information to defend users against themselves.

Reply:
Initial draft of request validation RFC: #11